node/src/store/block_ranges.rs

```diff
@@ -0,0 +1,977 @@
+            return true;
+        }
+
+        // `self` is subset of `other`
+        //
+        // self:    |--|
+        // other: |------|
+        if self.start() >= other.start() && self.end() <= other.end() {
+            return true;
+        }
+
+        // `self` is superset of `other`
+        //
+        // self:  |------|
+        // other:   |--|
+        if self.start() <= other.start() && self.end() >= other.end() {
+            return true;
+        }
+
+        false
+    }
+
+    /// Returns `true` if the whole range of `self` is on the left of `other`.
+    fn left_of(&self, other: &BlockRange) -> bool {
+        debug_assert!(self.validate().is_ok());
+        debug_assert!(other.validate().is_ok());
+        self.end() < other.start()
+    }
+}
+
+/// Represents possibly multiple non-overlapping, sorted ranges of header heights
+#[derive(Debug, Clone, PartialEq, Default, Serialize)]
+#[serde(transparent)]
+pub struct BlockRanges(SmallVec<[BlockRange; 2]>);
+
+/// Custom `Deserialize` that validates `BlockRanges`.
+impl<'de> serde::Deserialize<'de> for BlockRanges {
+    fn deserialize<D>(deserializer: D) -> Result<BlockRanges, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        let raw_ranges = SmallVec::<[RangeInclusive<u64>; 2]>::deserialize(deserializer)?;
+        let ranges = BlockRanges::from_vec(raw_ranges)
+            .map_err(|e| serde::de::Error::custom(e.to_string()))?;
+        Ok(ranges)
+    }
+}
+
+impl BlockRanges {
+    /// Create and empty `BlockRanges`.
+    pub fn new() -> BlockRanges {
+        BlockRanges(SmallVec::new())
+    }
+
+    pub(crate) fn from_vec(ranges: SmallVec<[BlockRange; 2]>) -> Result<Self> {
+        let mut prev: Option<&RangeInclusive<u64>> = None;
+
+        for range in &ranges {
+            range.validate()?;
+
+            if let Some(prev) = prev {
+                if range.start() <= prev.end() {
+                    return Err(BlockRangesError::UnsortedBlockRanges);
+                }
+            }
+
+            prev = Some(range);
+        }
+
+        Ok(BlockRanges(ranges))
+    }
+
+    /// Return whether `BlockRanges` contains provided height
+    pub fn contains(&self, height: u64) -> bool {
+        self.0.iter().any(|r| r.contains(&height))
+    }
+
+    /// Return whether range is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.iter().all(|r| r.is_empty())
+    }
+
+    /// Return highest height in the range
+    pub fn head(&self) -> Option<u64> {
+        self.0.last().map(|r| *r.end())
+    }
+
+    /// Return lowest height in the range
+    pub fn tail(&self) -> Option<u64> {
+        self.0.first().map(|r| *r.start())
+    }
+
+    /// Returns the start index and end index of the affected ranges.
+    ///
+    /// An affected range is the one that overlaps or touches `range`.
+    fn find_affected_ranges(&self, range: impl Borrow<BlockRange>) -> Option<(usize, usize)> {
```

Comment:
```suggestion
    fn find_neighbouring_ranges(&self, range: impl Borrow<BlockRange>) -> Option<(usize, usize)> {
```
or `adjacent`?

Reply:
it doesn't return only neighbors or adjacent. for example if we have these:

stored: 10..=15, 20..=25
input: 5..=30

It will return `(0,1)`, because those ranges are affected. The reason in this case is just overlapping.